### PR TITLE
Include signing metadata in merged manifest and push manifest as asset

### PR DESCRIPTION
**Pull request by jcagme**
Solves part of: https://github.com/dotnet/arcade/issues/5693

Once an official version of this package is published, it will need to be included in arcade with some changes in `PublishBuildAssets.proj`

Arcade build which consumed a test version with this change here: https://dnceng.visualstudio.com/internal/_build/results?buildId=712217&view=results

### CODE REVIEW

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs [hunk 1]
```diff
@@ -103,4 +106,69 @@ public class Blob
         [XmlAttribute(AttributeName = "NonShipping")]
         public bool NonShipping { get; set; }
     }
+
+    [XmlRoot(ElementName = "SigningInformation")]
+    public class SigningInformation
+    {
+        [XmlAttribute(AttributeName = "AzureDevOpsCollectionUri")]
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:113]: I don't have a lot of context in this signing work, but why are we duplicating the azdo build information in this section?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs [hunk 2]
```diff
@@ -103,4 +106,69 @@ public class Blob
...
+    {
+        [XmlAttribute(AttributeName = "File")]
+        public string File { get; set; }
+
+        [XmlAttribute(AttributeName = "CertificateName")]
+        public string CertificateName { get; set; }
+    }
+
+    [XmlRoot(ElementName = "ItemsToSign")]
+    public class ItemsToSign
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:156]: [nit] This seems to be a single ItemToSign rather than Items

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs [hunk 3]
```diff
@@ -103,4 +106,69 @@ public class Blob
...
+
+        [XmlAttribute(AttributeName = "CertificateName")]
+        public string CertificateName { get; set; }
+    }
+
+    [XmlRoot(ElementName = "FileSignInfo")]
+    public class FileSignInfo
+    {
+        [XmlAttribute(AttributeName = "File")]
+        public string File { get; set; }
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:149]: Is this supposed to be the full path? just the name? Would it make sense to have a File model that includes more information?

Comment (riarenas) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:149]: Same comment everywhere File is a string.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 4]
```diff
@@ -36,11 +38,19 @@ public class PushMetadataToBuildAssetRegistry : MSBuild.Task, ICancelableTask
 
         public string RepoRoot { get; set; }
 
+        public string AssetVersion { get; set; }
+
+        public string[] ManifestBuildData { get; set; }
+
+        public string PublishingVersion { get; set; } = "3";
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:45]: What are the implications of defaulting this to 3?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 5]
```diff
@@ -98,6 +130,10 @@ public async Task<bool> PushMetadataAsync(CancellationToken cancellationToken)
                     Log.LogMessage(MessageImportance.High, $"Metadata has been pushed. Build id in the Build Asset Registry is '{recordedBuild.Id}'");
                     Console.WriteLine($"##vso[build.addbuildtag]BAR ID - {recordedBuild.Id}");
 
+                    // Based on the in-memory merged manifest, create a physical XML file and
+                    // upload it to the BlobArtifacts folder
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:134]: Why BlobArtifacts instead of the AssetManifests container?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 6]
```diff
@@ -36,11 +38,19 @@ public class PushMetadataToBuildAssetRegistry : MSBuild.Task, ICancelableTask
...
+
+        public string[] ManifestBuildData { get; set; }
+
+        public string PublishingVersion { get; set; } = "3";
+
         [Output]
         public int BuildId { get; set; }
 
         private const string SearchPattern = "*.xml";
+        private const string MergeManifestFileName = "MergedManifest.xml";
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:51]: [nit] MergedManifestFileName

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 7]
```diff
@@ -437,6 +491,35 @@ private BuildData MergeBuildManifests(List<BuildData> buildsMetadata)
             return mergedBuild;
         }
 
+        private SigningInformation MergeSigningInfo(List<SigningInformation> signignInformation)
+        {
+            SigningInformation mergedInfo = signignInformation[0];
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:496]: Can this list ever be empty so we need to guard before this access?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 8]
```diff
@@ -437,6 +491,35 @@ private BuildData MergeBuildManifests(List<BuildData> buildsMetadata)
...
+
+            for (int i = 1; i < signignInformation.Count; i++)
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:498]: Any reason not to foreach for readability?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 9]
```diff
@@ -437,6 +491,35 @@ private BuildData MergeBuildManifests(List<BuildData> buildsMetadata)
             return mergedBuild;
         }
 
+        private SigningInformation MergeSigningInfo(List<SigningInformation> signignInformation)
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:494]: [nit] typo: signingInformation

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 10]
```diff
@@ -495,5 +578,186 @@ private void LookupForMatchingGitHubRepository(BuildData mergedBuild)
...
+
+                if (asset != null)
+                {
+                    AssetVersion = asset.Version;
+                }
+            }
+
+            AssetData assetData = new AssetData(true)
+            {
+                Locations = (location == null) ? null : ImmutableList.Create(new AssetLocationData(LocationType.Container)
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:606]: Is a null location here handled correctly everywhere? Is it safer to return an empty list?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 11]
```diff
@@ -495,5 +578,186 @@ private void LookupForMatchingGitHubRepository(BuildData mergedBuild)
...
+
+            if (finalSigningInfo != null)
+            {
+                buildModelXml.Add(SigningInfoToXml(finalSigningInfo));
+            }
+
+            File.WriteAllText(mergedManifestPath, buildModelXml.ToString());
+
+            Log.LogMessage(MessageImportance.High,
+                        $"##vso[artifact.upload containerfolder=BlobArtifacts;artifactname=BlobArtifacts]{mergedManifestPath}");
```
Comment (riarenas) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:702]: Should this value be extracted as a constant? Is the AssetManifests container more appropriate?

**Comment by riarenas** [comment]
I don't have a lot of context in the merged manifest so maybe someone else with more context in the work should also review., Left some comments about duplication and where the merged manifest will be persisted. 

Would also really like to see tests for the new functionality.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 1 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:113]: @jonfortescue, do you know? I think what @riarenas means is that we already have build info in the `Build` node

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 2 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:156]: That's the name of the element i.e. `<ItemsToSign File="Microsoft.DotNet.Build.Tasks.Feed.5.0.0-ci.nupkg" />`. @jonfortescue fyi

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 3 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:149]: It juts represents the name of the file we'll sign

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 4 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:45]: Talked to @mmitche and he suggesting using 3 since this would be the "new" form of publishing since we'll include the manifest as a blob item compared to previous versions where this is not included. Other than that I don't think there are other implications

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 5 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:134]: Manifest is already been pushed to `AssetManifest` but since now we are treating it as a Blob, we need to push it, also, to the `BlobArtifacts` container

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 7 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:496]: Yeah, the check is been done in line 96 `if (signingInformation.Any())`

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 5 above)
Comment (riarenas) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:134]: I looked at the test build, and I could only find the asset in the BlobArtifacts, and not in the AssetManifests container, so I think it's only being uploaded there?

Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:134]: Oh, what I meant was that `Windows_NT-AnyCPU.xml` in the test build is published to AssetManifest. AssetManifets is the base folder we use to generate the MergedManifest so while we could push it to AssetManifests it won't have any effect.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 8 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:498]: Mainly to avoid logic to check for the first iteration of the loop (line 496) and the rest

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 5 above)
Comment (riarenas) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:134]: I'm just wondering if the right thing to do is to always publish it to assetManifests, and then change the publishing code to always look for a mergedManifest.xml there and upload to blob storage from there, but that's probably not worth it.

Comment (riarenas) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:134]: So I think putting it in blobArtifacts is fine as it conveys more appropriately that it should be published as a blob.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 8 above)
Comment (riarenas) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:498]: I just think this can be rewritten to guard against bad callers that didn't explicitly check for an empty signingInformation before calling the method.
Initialize with null, iterate with a foreach. Return ends up being null if there is no info.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 10 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:606]: `AssetData` is what is POSTed to BAR and null (for empty) is expected.

```json
{
      "id": 4782687,
      "name": "assets/symbols/Microsoft.DotNet.AsmDiff.5.0.0-beta.20351.8.symbols.nupkg",
      "version": "5.0.0-beta.20351.8",
      "buildId": 56549,
      "nonShipping": true,
      "locations": null
    },
```

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 11 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:702]: This won't be reused so I don't want to declare a variable for a single use. Also, as previously discussed, this will end up in BlobArtifacts

**Comment by riarenas** [comment]
Also CC @epananth in case she sees anything here that might interact weirdly with the publishing work.

**Comment by jcagme** [comment]
This is the test build after the changes: https://dnceng.visualstudio.com/internal/_build/results?buildId=718291&view=results. @mmitche could you PTAL?

**Comment by jcagme** [comment]
@mmitche could you PTAL?

**Comment by mmitche** [comment]
Taking a look now.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileExtensionSignInfoComparer.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+using System;
```
Comment (mmitche) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileExtensionSignInfoComparer.cs:1]: NIt: missing header

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileSignInfoComparer.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
+using System;
```
Comment (mmitche) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileSignInfoComparer.cs:1]: nit: header

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileSignInfoComparer.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+{
+    /// <summary>
+    ///     Compares FileSignInfo objects based on file and certificate name
+    /// </summary>
+    public class FileSignInfoComparer : IEqualityComparer<FileSignInfo>
+    {
+        public bool Equals(FileSignInfo x, FileSignInfo y)
+        {
+            return x.CertificateName.Equals(y.CertificateName, StringComparison.OrdinalIgnoreCase) &&
+                x.File.Equals(y.File, StringComparison.OrdinalIgnoreCase);
```
Comment (mmitche) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileSignInfoComparer.cs:14]: This file is a sub-path?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/ItemsToSignComparer.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+namespace Microsoft.DotNet.Maestro.Tasks
+{
+    /// <summary>
+    ///     Compares ItemsToSign objects based on file
+    /// </summary>
+    public class ItemsToSignComparer : IEqualityComparer<ItemsToSign>
+    {
+        public bool Equals(ItemsToSign x, ItemsToSign y)
+        {
+            return x.File.Equals(y.File, StringComparison.OrdinalIgnoreCase);
```
Comment (mmitche) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/ItemsToSignComparer.cs:13]: This file is a subpath I assume? As in we wouldn't see any weirdness file layouts were different on two different machines?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 1 above)
Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:113]: Yeah I don't think this would be needed?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 3 above)
Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:149]: Okay so we assume that foo.dll is basically the same everywhere

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 2 above)
Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:156]: I'm with @riarenas on this. It seems to only represent a single item to sign.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileSignInfoComparer.cs (hunk 14 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/FileSignInfoComparer.cs:14]: Yeah it is just the file name with no path at all. `<FileSignInfo File="YamlDotNet.dll" CertificateName="3PartySHA2" />`

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/ItemsToSignComparer.cs (hunk 15 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/ItemsToSignComparer.cs:13]: Yeah it is just the file name with no path at all.  `<ItemsToSign File="Microsoft.Cci.Extensions.5.0.0-ci.nupkg" />`

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 3 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:149]: I think that to be the case. @jonfortescue can you confirm?

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 4 above)
Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:45]: The code generating the manifest in each build leg: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Build.Tasks.Feed/src/GenerateBuildManifest.cs#L75

Is defaulting to the Legacy version right now. I think if only if the input versions were non-legacy would signing information end up in the manifest.  So perhaps keying off the input manifest versions would be the right thing to do. If those are all 3 (they shouldn't differ, and that should be checked here), then we would trigger the new merged manifest w/signing info behavior.

Does that make sense?

Basically what we're saying is the v1 is legacy (3.x), v2 is what we currently have, and v3 is the future. Any kind of breaking change (like this one), should go into 3. Using 3 requires that we have all v3 input manifests. Prior to this change, we would never had a merged manifest (the merging was implicit during publishing), so if all inputs were 3 then we're on 3. Now we're adding a new merged manifest (allowing us to ignore the other non-merged ones). It's version should match the publishing version of all input manifests.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs (hunk 5 above)
Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:129]: I think either approach works. This is probably fine.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 16]
```diff
@@ -495,5 +581,186 @@ private void LookupForMatchingGitHubRepository(BuildData mergedBuild)
...
+            BuildModel buildModel = new BuildModel(
+                        new BuildIdentity
+                        {
+                            Attributes = GetNamedProperties(ManifestBuildData),
+                            Name = GetAzDevRepository(),
+                            BuildId = GetAzDevBuildNumber(),
+                            Branch = GetAzDevBranch(),
+                            Commit = GetAzDevCommit(),
+                            IsStable = IsStableBuild.ToString(),
+                            PublishingVersion = PublishingVersion
```
Comment (mmitche) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:663]: This should be the merge of the input publishing versions.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 2 above)
Comment (jonfortescue) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:156]: Ultimately this is based on the [Build Manifest (signing)](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/src/GenerateAzureDevOpsBuildManifest.cs#L15) that @chcosta made in the Arcade SDK. This model is in parity with the [SigningInformationModel](https://github.com/dotnet/arcade/pull/5517/files#diff-a3a076be8faf428780c5bb3c62d363c3R17) introduced in my PR.

I can change it in all places in my PR but that's where this ultimately derives from.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 3 above)
Comment (jonfortescue) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:149]: Yep, exactly, as per [Signing.props in arcade](https://github.com/dotnet/arcade/blob/master/eng/Signing.props#L7-L29).

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 1 above)
Comment (jonfortescue) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:113]: This comes from [here](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj#L252). Tbh I'm not entirely sure what it does; I based my models on @chcosta's signing manifest which was designed to be used in the release pipeline. I'm not entirely sure off the top of my head what the intent was.

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs (hunk 2 above)
Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/Models/Manifest.cs:159]: [`sign.proj`](https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Sign.proj#L46) expects `ItemsToSign` so I assume that's why we went this naming even though we are including a single item. Like this, we can just use the metadata as is without having to do any change

File: src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs [hunk 17]
```diff
@@ -295,6 +299,18 @@ private string GetVersion(string assetId)
                 {
                     var manifest = (Manifest)xmlSerializer.Deserialize(stream);
 
+                    if (publishingVersion == -1)
+                    {
+                        publishingVersion = manifest.PublishingVersion;
```
Comment (mmitche) [inline src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:304]: Legacy manifests may not have the publishing version (I think master sets it appropriately right now, but 3.x may not have that change). I would omit the attribute altogether in the merged manifest if it doesn't exist in the input manifests.

Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:304]: Well if it is not set legacy is assumed which won't create a MergedManifest so there's no need to not include it

Comment (jcagme) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:304]: Also, this same code will work since `manifest.PublishingVersion` will end up being 0 so the same logic would work

Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:304]: The publishing infra will treat 0 as an error:

https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs#L213-L232

Comment (mmitche) [reply src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:304]: Also see: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifest.cs#L168-L188
